src/python/dxpy/bindings/apollo/expression_matrix_transformation.py

```diff
@@ -0,0 +1,63 @@
+from copy import deepcopy
+
+def expression_transform(results_list):
+    """
+    results_list: list of dictionaries, each of the form
+    {
+        "feature_id":<feature_id>,
+        "sample_id":<sample_id>,
+        "expression":<expression>
+    }
+    Additional key:value pairs are possible, but are not included in the transformed output
+
+    How it works:
+    First, create a dictionary where each key is a sample id, and the values are the feature_id:expression pairs
+    Then, convert it to a list of dictionaries (the format that output handling is expecting)
+    Each item in the list is a dictionary containing one "sample_id":sample_id pair, and
+    multiple feature_id:expression pairs
+
+    Returns:
+    A list of dictionaries, each of the form:
+    {
+        "sample_id":<sample_id>,
+        <feature_id_1>:<expression_1>
+        <feature_id_2>:<expression_2>
+        ...
+        <feature_id_n>:<expression_n>
+    }
+    """
+    transformed_dict = {}
+    # create a dict of the form
+    # {
+    #   <sample_id>:{
+    #       <feature_id_1>:<expression_1>,
+    #       <feature_id_2>:<expression_2>,
+    #        etc.
+    #   }
+    # 
+    # }
+    for entry in results_list:
+        if entry["sample_id"] not in transformed_dict:
+            transformed_dict[entry["sample_id"]] = {entry["feature_id"]:entry["expression"]}
+        else:
+            transformed_dict[entry["sample_id"]][entry["feature_id"]] = entry["expression"]
+
+    # Modify the above dictionary into a list of dictionaries, essentially moving the key of each
+    # dict into the dict itself
+    dict_list = []
+    for sample in transformed_dict:
+        samp_row = deepcopy(transformed_dict[sample])
+        samp_row["sample_id"] = sample
+        dict_list.append(samp_row)
+    
+    # Get the column names that the output writer will use to generate the table
+    # Can't use set here because we want the order of the columns to be deterministic
+    colnames = []
+    for row in results_list:
```

Comment:
I think if you start with a `set` of unique `feature_id`s then you might be able to run the second for loop right within the first one, and then `deepcopy` might not be necessary anymore. But this will probably require some changes in the current code in both loops.

Reply:
One way would be to just start filling in `sample_id`s right here immediately after `for row in results_list`. Then the second nested for loop would just fill in the `value` based on the current, matched `feature_id`.